pkg/systemlogmonitor/log_monitor.go

```diff
@@ -196,6 +212,19 @@ func initialConditions(defaults []types.Condition) []types.Condition {
 	return conditions
 }
 
+func initialProblemGauges(initialConditions []types.Condition) {
+	for _, condition := range initialConditions {
+		if condition.Status != types.True {
```

Comment:
In `initialConditions`, the initial ones are all set to false, so this function becomes noops, right?

I think the purpose of this function is to create problem_gauges that have conditions being false, right (the so called default conditions)?

Reply:
> In initialConditions, the initial ones are all set to false, so this function becomes noops, right?

Correct. However, this function is still useful **if** some condition has initial value set to True.

> I think the purpose of this function is to create problem_gauges that have conditions being false

That was the case for k8s conditions. However, I don't think the same UI looks good with Prometheus. Mostly because a metric with different labels are consider as multiple metrics in Prometheus, and the UI looks ugly.

For example, if I keep the code, and skip when the `condition.Status != True`, then this is what I get (which look pretty OK):
```
# HELP problem_gauge Whether a specific type of problem is affecting the node or not.
# TYPE problem_gauge gauge
problem_gauge{reason="DockerHung",type="KernelDeadlock"} 1
```

And if I remove the skip logic, then the output looks like this:
```
# HELP problem_gauge Whether a specific type of problem is affecting the node or not.
# TYPE problem_gauge gauge
problem_gauge{reason="DockerHung",type="KernelDeadlock"} 1
problem_gauge{reason="FilesystemIsNotReadOnly",type="ReadonlyFilesystem"} 0
problem_gauge{reason="KernelHasNoDeadlock",type="KernelDeadlock"} 0
```

Basically, the initial states `problem_gauge{reason="KernelHasNoDeadlock",type="KernelDeadlock"} 0`  and `problem_gauge{reason="FilesystemIsNotReadOnly",type="ReadonlyFilesystem"} 0` will show up there all the time. That seems a bit ugly and misleading.

Reply:
Hmm, I think we need to clarify what does `0` and `1` mean for the problem gauge.

With your current implementation:
- For `DockerHung`, `0` means no such condition, `1` means it has such condition.
- However, for `KernelHasNoDeadlock`, `0` now means it has such condition (the condition is that "there is no deadlock". And the value will never be `1`.

The above looks inconsistent. IMO, the value `0` and `1` should just mean `false` and `true`. So the initial state for `KernelHasNoDeadlock` should be:
```
problem_gauge{reason="KernelHasNoDeadlock",type="KernelDeadlock"} 1
```

Later when we have `DockerHung`, it should become:
```
problem_gauge{reason="DockerHung",type="KernelDeadlock"} 1
problem_gauge{reason="KernelHasNoDeadlock",type="KernelDeadlock"} 0
```

Reply:
1. 
> However, for KernelHasNoDeadlock, 0 now means it has such condition (the condition is that "there is no deadlock". And the value will never be 1.

> The above looks inconsistent. IMO, the value 0 and 1 should just mean false and true. So the initial state for KernelHasNoDeadlock should be:

I guess there is a misunderstanding. `KernelHasNoDeadlock` is actually not the permanent problem (i.e. not the `condition`), it is the `reason` field of the `KernelDeadlock` condition.

So `0` or `1` does not indicate whether `KernelHasNoDeadlock` is happening or not; it indicates whether `KernelDeadlock` is happening or not.
`reason=xxx` field explain why is `KernelDeadlock` happening (e.g. `DockerHung`), or why is it not happening (e.g. `KernelHasNoDeadlock`, in which case I think is very confusing).

2. 

First I want to say that both `problem_gauge{reason="KernelHasNoDeadlock",type="KernelDeadlock"} 0` and `problem_gauge{reason="KernelHasNoDeadlock",type="KernelDeadlock"} 1` is very confusing and misleading.

What we are trying to say is, "the node is not suffering from KernelDeadlock problems", and the reason of this is "the kernel has no deadlock".

I think the best UI is to not show the "KernelHasNoDeadlock" field at all.

3.

> I think we need to clarify what does 0 and 1 mean for the problem gauge.

In my opinion, for problem gauge, `0` and `1` should indicate "whether a specific type of problem is affecting the node or not" (the quote is the description of the `problem_gauge` Prometheus metric).

So when the value is `0`, it should mean "this type of problem is NOT affecting the node"; when the value is `1`, it should mean "this type of problem is affecting the node".

4.

I think showing `problem_gauge{reason="KernelHasNoDeadlock",type="KernelDeadlock"}` only confuses the user, and has no real benefit. i.e. If a permanent problem is not happening, the user don't care why it's not happening. The `reason` is always gonna be a very dumb thing to say, like "Kernel is not deadlocking", "OOM is not happening". I think it only confuses the user.

The right user experience should be, they simply build a Prometheus query, and see whether "problem_gauge{type="KernelDeadlock"}" is 0 (using `addition` to aggregate all sub-metrics like `problem_gauge{type="KernelDeadlock", reason="DockerHung"}` or `problem_gauge{type="KernelDeadlock", reason="ContainerdHung"}` ).

If it's 0, then the problem is not affecting the node; if it's 1, then the problem is affecting the node.

And if it's 1, the 1-valued sub-metric should have a `reason` field describing what caused the problem.

WDYT? Is this UX reasonable?

Reply:
Your points make sense. We do not need `KernelHasNoDeadlock`.

I think what I tried to get is to have some initial value for the problems (can be 0). The metrics that I have seen so far have values starting from the beginning, like disk io, cpu, memory, etc. So you can see a line starting from the beginning of the life of the node and lasting till the end of the life of the node.

For problem gauges, it seems to me that metrics start to have values only when some problems occur. I am not sure if that's going to be an issue or not. For example, for most of the time, there is no metric for problem_gauge and problem_counter when no issue occurs. This makes me feel inconsistent with other metrics, where we always have some values when a node is alive.

What in my mind is more like:

When node is booted, we have
```
problem_gauge{reason="DockerHung",type="KernelDeadlock"} 0
problem_gauge{reason="AUFSUmountHung",type="KernelDeadlock"} 0
```

Then when we have AUFSUmountHung, we get
```
problem_gauge{reason="DockerHung",type="KernelDeadlock"} 0
problem_gauge{reason="AUFSUmountHung",type="KernelDeadlock"} 1
```

Then we have DockerHung, we get
```
problem_gauge{reason="DockerHung",type="KernelDeadlock"} 1
problem_gauge{reason="AUFSUmountHung",type="KernelDeadlock"} 0
```

Similarly for temporary problems, the initial values are like
```
problem_counter{reason="TaskHung"} 0
```

Reply:
Ah I see. I like that. I will implement this behavior today. Thanks!

Reply:
Done.